File: thinking-categorically/applicative-functor/arrow-functor/arrow-functor.cs
using System;

class Program
{
    static void Main()
    {
        // map: (A → B) → (R → A) → (R → B)
        Func<Func<A, B>, Func<R, A>, Func<R, B>> Map<A, B, R>() =>
            (f, g) => x => f(g(x));

        // Functions
        Func<int, int> add1 = x => x + 1;
        Func<int, string> add2 = x => $"{x} + 2";

        // Compose functions using map
        var result = Map<int, string, int>()(add2, add1);

        Console.WriteLine(result(5)); // prints: "6 + 2"
    }
}

File: thinking-categorically/applicative-functor/arrow-functor/arrow-functor.ts
type Arrow<A, B> = (a: A) => B;
type F<R, A> = Arrow<R, A>;

const f: Arrow<number, string> = x => `${x} + 2`;   // A -> B
const g: Arrow<number, number> = x => x + 1;        // R -> A

const composed: Arrow<number, string> = (r: number) => f(g(r)); // R -> B

const fmap =
  <R>() =>
  <A, B>(f: Arrow<A, B>) =>
  (g: Arrow<R, A>): Arrow<R, B> =>
    (r: R) => f(g(r));

const add1: Arrow<number, number> = x => x + 1;
const add2: Arrow<number, string> = x => `${x} + 2`;

const result = fmap<number>()(add2)(add1);

console.log(result(5)); // '6 + 2'
